[1단계 - 자동차 경주 구현] 트레(이충렬) 미션 제출합니다.

takoyakimchi opened the pull request:
[1단계 - 자동차 경주 구현] 트레(이충렬) 미션 제출합니다.

takoyakimchi committed:
docs: 기능 명세서 작성

takoyakimchi committed:
feat(Car): Car 도메인 생성

takoyakimchi committed:
feat(view): 사용자로부터 문자열 입력받는 함수 생성

takoyakimchi committed:
feat(validation): 자동차 이름 규칙 판별 로직 추가

takoyakimchi committed:
feat(validation): 자동차 이름 규칙 중복 판별 로직 추가

takoyakimchi committed:
feat(validation): 시도 횟수 검증 로직 추가

takoyakimchi committed:
feat(validation): 시도 횟수 검증 로직 추가

takoyakimchi committed:
style(validation): 메서드 위치 조정

takoyakimchi committed:
feat(validation): 차량 대수 검증 로직 추가

takoyakimchi committed:
feat(conversion): 사용자 입력 문자열 -> List<Car> 변환 로직 추가

takoyakimchi committed:
feat(conversion): 시도 횟수 문자열 -> int 변환 로직 추가

takoyakimchi committed:
feat(view): 단순 텍스트 출력 로직 추가

takoyakimchi committed:
refactor(Car): getter 추가

takoyakimchi committed:
feat(view): 라운드 별 실행 결과 출력 로직 추가

takoyakimchi committed:
feat(view): 우승자 출력 로직 추가

takoyakimchi committed:
feat(service): 전진 여부 판단 로직 추가

takoyakimchi committed:
feat(service): 자동차 전진 로직 추가

takoyakimchi committed:
feat(service): 자동차 전진 로직 추가

takoyakimchi committed:
refactor(service): 전진 기준 숫자 및 랜덤 범위 상수화

takoyakimchi committed:
feat(service): 우승자 판단 로직 추가

takoyakimchi committed:
feat(controller): 메인 로직을 실행하는 컨트롤러 추가

takoyakimchi committed:
feat(main): 메인 함수 추가

takoyakimchi committed:
docs: TODO 작성

takoyakimchi committed:
refactor(service): moveForward 메서드 삭제 -> playOneRound 메서드로 역할 이동

playOneRound 메서드는 한 라운드에 대해서 모든 자동차의 전진 여부를 판단하고 전진 or 정지하도록 함

takoyakimchi committed:
refactor(service): 상수에 final 적용

takoyakimchi committed:
refactor(view): 각 라운드 끝날 때마다 new line 출력하는 로직을 controller -> view로 이동

takoyakimchi committed:
style(controller): 불필요한 TODO 삭제

takoyakimchi committed:
refactor(controller): 자동차 이름, 시도 횟수 입력받는 기능 메서드화

takoyakimchi committed:
feat(view): 값은 반복해서 입력받고 예외가 발생하면 예외 처리하는 IterativeReader 메서드 생성

takoyakimchi committed:
refactor(controller): 잘못된 값이 입력되었을 때 재입력받는 기능 적용

takoyakimchi committed:
style(view): 해결된 TODO 삭제

takoyakimchi committed:
refactor(view): 불필요하게 String 객체 생성하지 않도록 변경

takoyakimchi committed:
refactor: 상수를 별도의 클래스로 분리

takoyakimchi committed:
feat(test): Validator 클래스에서 차 이름 테스트 추가

takoyakimchi committed:
feat(test): Validator 클래스에서 시도 횟수 테스트 추가

ValidatorTest 완성

takoyakimchi committed:
feat(test): Service 우승자 판별(정상 입력) 테스트 작성

takoyakimchi committed:
style(test): ServiceTest에서 DisplayName 변경

takoyakimchi committed:
feat(test): Converter 테스트 추가

Converter에서 toCars() 및 toRound()에 대한 테스트 추가

takoyakimchi committed:
refactor(test): ValidatorTest에서 라운드 테스트(예외 입력) 테스트 케이스 추가

takoyakimchi committed:
style(controller): 메서드 위치 변경

takoyakimchi committed:
style(service): 메서드 위치 변경 - public을 위로

takoyakimchi committed:
style(validation): 메서드 위치 변경 - public을 위로, private 메서드 호출되는 순서에 따라 순서 변경

takoyakimchi committed:
style(view): IterativeReader.errorHandledRead -> readUntilNoError 메서드명 변경

takoyakimchi committed:
style(const): 상수 이름을 메시지와 동일한 의미를 갖도록 변경

takoyakimchi committed:
refactor(conversion): 자동차 이름 입력 시 comma 앞뒤 공백 제거

takoyakimchi committed:
refactor(test): 자동차 이름 입력 시 구분자 전후 공백 제거 테스트

takoyakimchi committed:
style: 불필요한 주석 및 import 제거

pkeugine commented on src/main/java/constant/Messages.java:
```diff
@@ -0,0 +1,9 @@
+
+public class Messages {
+
+    public static final String INPUT_CAR_NAMES = "경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).";
+    public static final String INPUT_ROUND = "시도할 회수는 몇회인가요?";
+    public static final String GAME_RESULT = "실행 결과";
+    public static final String WIN_THE_GAME = "가 최종 우승했습니다.";
```
view 관련 메시지를 별도의 Messages 클래스에 모으면 어떤 장단점이 있을까요?

takoyakimchi replied:
실제 서비스에서 생각해 보았을 땐 국제화 등의 처리를 하기가 편해질 것 같다는 생각이 들어
연습 삼아 시도해 보았습니다!

하지만 이렇게 하니 실제 print하는 클래스에서 어떤 게 print 되는지 한 번에 알기가 어렵고
print해야 할 메시지가 많아지는 경우 클래스가 커질 수도 있을 것 같습니다

pkeugine replied:
오 국제화! 국제화처럼 메시지에 따로 처리할 로직이 있다면 확실히 모아둘 필요가 있겠죠 👍 

트레가 언급해주신 것처럼 막상 메시지를 사용하는 쪽에서 그 내용을 한 번에 알 수 없고,
예외가 많아질수록 관리할 때 신경을 더 많이 써야한다는 단점이 있겠네요
예리한 답변입니다 :)

pkeugine commented on src/main/java/constant/Numbers.java:
```diff
@@ -0,0 +1,12 @@
+package constant;
+
+public class Numbers {
+    public static final int RANDOM_NUMBER_RANGE = 10;
+    public static final int MOVE_BOUNDARY_NUMBER = 4;
+
+    public static final int MIN_ROUND = 1;
+    public static final int MAX_ROUND = 100;
+
+    public static final int MIN_CAR_LENGTH = 1;
+    public static final int MAX_CAR_LENGTH = 100;
+}
```
Messages 와 유사하게 magic number 들을 한 곳에 모으셨군요!
수정될 수 있는 숫자를 한 곳에서 관리하기 위함일까요?
트레가 이 구조를 선택한 이유가 궁금해요 :)

takoyakimchi replied:
이 숫자들이 애플리케이션의 strategy라고 생각이 들었고 
이러한 것들을 한 번에 처리할 수 있는 클래스가 있으면 좋을 것 같아서
한 곳에 모아서 처리하게 되었습니다!

하지만 지금처럼 애플리케이션의 모든 상수들을 한 클래스에서 정의하는 방식은 한계가 있다고 생각합니다
의미에 따라서 클래스를 분리하는 것이 좋을지, 아니면 상수를 사용하는 클래스에서 상수를 정의하는 것이 좋을지
궁금합니다!

pkeugine replied:
트레가 앞서 에러 메시지를 한 곳에 묶었을 때 국제화와 같은 장점을 언급해주신 것처럼
상수를 한 곳에 묶었다면 그만한 장점이 있어야할 것 같아요

사용처에 상수를 정의하면 코드를 유지보수 할 때 편리해져서 좋다고 생각해요
로직을 이해하고 변경할 땐 아무래도 사용하는 곳에 가까이 상수를 두는 편이 유지보수하기 편하지 않나 싶어요
그 값을 바로 알 수도 있고, 이름이라던지 그 구조를 수정할 때 변경점도 덜 잡힐테니까요 🤔

pkeugine commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,39 @@
+
+        List<String> winnerNames = Service.getWinnerNames(cars);
+        OutputView.printWinners(winnerNames);
+    }
```
Controller 클래스가 static 하게 만들어져 있네요
이런 구조를 선택하신 이유가 궁금합니다!

이 구조라면, `start()` 메서드를 호출하는 Main 클래스가 대신해서 이 역할을 담당할 수도 있을 것 같네요 🤔

takoyakimchi replied:
이 부분은 사실 MVC 개념을 숙지하지 못한 채 막연하게 구현했습니다 😭😭
말씀주신 대로 Main으로 옮겨서 진행하는 것이 저도 좋다고 생각합니다!

pkeugine commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,39 @@
+        OutputView.printWinners(winnerNames);
+    }
+
+    private static List<Car> getCars() {
+        OutputView.printCarNames();
+        String rawCarNames = InputView.read();
+        return Converter.toCars(rawCarNames);
+    }
+
+    private static int getRound() {
+        OutputView.printRound();
+        String rawRound = InputView.read();
+        return Converter.toRound(rawRound);
+    }
```
제 개인적인 방식인데,
getter 에서 get 이라는 표현을 쓰다보니
getter 가 아닌 메서드에서는 웬만하면 get 이라는 표현을 쓰지 않으려고 스스로 노력해요

get 대신 find 를 주로 사용하는데요,
이건 그냥 제 개인적인 생각이니 참고용으로만 봐주세요 :)

takoyakimchi replied:
메서드 이름을 짓는 과정에서 같은 고민을 했었는데
굉장히 좋은 방식이네요!
이 부분도 수정하겠습니다.

pkeugine commented on src/main/java/conversion/Converter.java:
```diff
@@ -0,0 +1,35 @@
+            cars.add(new Car(carName));
+        }
+        return cars;
+    }
```
rawCarNames 를 자동차 객체인 cars 로 바꿔주는 로직은 언뜻보면 Converter 라는 static 한 유틸 클래스에 적합해보여요
하지만 이렇게 넣다보면 거의 모든 전환 로직이 Converter 에 만들어질거고,
- Converter 가 엄청 방대해지거나
- XxxConverter, YyyConverter, ZzzConverter 처럼 유틸 클래스가 객체를 대신하는 애플리케이션이 만들어질거에요

현재 트레가 만드신 Car 객체를 보면 getter 와 move 로 이루어진 단순한 도메인이에요.
한 번 이러한 변환로직을 도메인 객체에 넣어보면 어떨까요?

takoyakimchi replied:
동의합니다! 도메인 객체에 변환 로직 추가해 보도록 하겠습니다.

pkeugine commented on src/main/java/conversion/Converter.java:
```diff
@@ -0,0 +1,35 @@
+        Validator.validateRound(round);
+        return round;
+    }
+}
```
round 를 생성하는 로직이 깔끔하고 잘 만들어져 있어요.
그런데 자세히 보면 rawRound 가 integer 인지 검증하는 로직과 `Validator.validateRound` 가 공존하고 있어요.

그렇다면 rawRound 가 integer 인지 검증하는 것은 `validateRound` 와는 별개일까요?

takoyakimchi replied:
1. 입력 자체에 대한 validation
2. 유효 범위에 대한 validation

이렇게 나누고 싶었는데, `validateRound`라는 의미상으로 봤을 때는 별개가 아닌 것 같습니다!
메서드의 네이밍을 수정하거나 `validateRound` 한 곳에 모으는 작업이 필요할 것 같습니다.

pkeugine commented on src/main/java/conversion/Converter.java:
```diff
@@ -0,0 +1,35 @@
+        Validator.validateRound(round);
+        return round;
+    }
+}
```
클래스 [파일마다 새로운 line 을 넣어주는게 어떨까요?](https://gist.github.com/camh-/1bebfcff1b0f814e9b191edc60d5206b)
현재 PR 에서 변경점을 보면 아래와 같이 빨간 아이콘이 뜨거든요 : 
<img width="90" alt="Screenshot 2024-02-16 at 7 14 41 AM" src="https://github.com/woowacourse/java-racingcar/assets/48251668/5c07031c-9dd5-4bc4-8ac4-237510ad525d">


intelliJ 설정에서 개발자가 변경하는 파일마다 자동으로 파일 마지막에 new line 을 넣어주는 기능이 있어요. 이를 사용해보셔도 좋아요!

new line 이 사용되는 이유가 궁금하다면? [[클릭!](https://stackoverflow.com/questions/5813311/whats-the-significance-of-the-no-newline-at-end-of-file-log)]
intelliJ 에서 파일이 저장될 때 마지막 new line 을 넣어주는 방법이 궁금하다면? [[클릭!](https://stackoverflow.com/questions/16761227/how-to-make-intellij-idea-insert-a-new-line-at-every-end-of-file)]

takoyakimchi replied:
IDE에서 Reformat code 단축키를 사용하면 항상 마지막에 new line이 삽입되어서 의문이었는데
궁금했던 점이 어느 정도 해소되었습니다! 감사합니다.

1. POSIX 명세: 행의 끝은 개행이다. 텍스트 파일은 행의 집합이며 반드시 개행으로 끝난다.
2. 마지막 개행 문자가 없으면 일부 환경에서는 예기치 못한 문제가 발생할 수 있다.
3. 마지막 개행 문자가 없으면 파일 간 차이를 알기 어렵다.

pkeugine commented on src/main/java/domain/Car.java:
```diff
@@ -0,0 +1,24 @@
+
+public class Car {
+
+    private final String name;
```
객체 생성과 동시에 변경될 일이 없는 값 불변처리! 좋네요 👍

takoyakimchi replied:
공통 피드백 강의에서 네오가 `final`을 적극적으로 사용하시는 것을 보고 배웠습니다!
감사합니다!

pkeugine commented on src/main/java/domain/Car.java:
```diff
@@ -0,0 +1,24 @@
+        this.score = 0;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public int getScore() {
+        return score;
+    }
+
+    public void move() {
+        this.score++;
+    }
```
이번에 페어와 함께 개발하시면서 서로 정한 컨벤션이 있을까요?
클래스의 메서드 순서도 컨벤션이 될 수 있어요 :
생성자, 부생성자, public 메서드, private 메서드, getter, toString, compareTo, equals, hashCode 등

이 순서 컨벤션을 맞추면 코드를 유지보수할 때 굉장히 큰 도움이 되더라구요!

takoyakimchi replied:
상세하게는 정하지 못했지만 페어가 작성한 코드에서 private 메서드가 public 메서드보다 위에 위치하는 것을 보고

생성자 -> getter(필드 순서대로) -> public 메서드 -> private 메서드

이 순서에 맞게 페어에게 해당 내용 전달 후 제가 수정했던 적이 있습니다!

pkeugine commented on src/main/java/service/Service.java:
```diff
@@ -0,0 +1,37 @@
+import domain.Car;
+import java.util.List;
+
+public class Service {
```
벌써부터 Service 를 만들다니...!!
만들어보는 것도 좋지만, 앞서 남긴 리뷰에 이어서 제 생각을 남겨볼게요:

Controller 와 Service 가 모두 static 하게 만들어진다면
자동차 경주 수준의 로직을 봤을 때 과한 클래스 분리로 느껴지기도 해요
Main 클래스에서 다루거나, 일부는 domain 에게 책임을 위임할 수 있는 로직들이거든요

그런데 Service 안에서는 domain 만 사용해서 로직을 구성한건 👍

takoyakimchi replied:
과한 클래스 분리!!! 동의합니다
과제를 진행하면서도 클래스를 분리해서 구현해야겠다는 막연한 생각 때문에
오히려 복잡하게 느껴지기도 했습니다.
개선해 보도록 하겠습니다!

pkeugine commented on src/main/java/service/Service.java:
```diff
@@ -0,0 +1,37 @@
+            .max()
+            .orElseThrow(RuntimeException::new);
+
+        List<String> winnerNames = cars.stream()
+            .filter(car -> car.getScore() == maxScore)
+            .map(Car::getName)
+            .toList();
+
+        return winnerNames;
```
```suggestion
        return cars.stream()
            .filter(car -> car.getScore() == maxScore)
            .map(Car::getName)
            .toList();
```
곧바로 return 할 객체를 만들었다면 지역 변수에 담지 않고 바로 반환할 수도 있어요!

takoyakimchi replied:
의미적으로 구분하기 위해서 IDE의 잔소리를 무시하고
꿋꿋하게 변수를 만드는 스타일로 작성을 했던 것 같습니다!
불필요하게 변수를 만든다면 (현재는 큰 문제 없겠지만) 메모리 낭비 문제가 발생할까요?

takoyakimchi replied:
이 부분은 

1. 최대 점수 찾기
2. 최대 점수를 갖는 Car 찾기

두 개의 메서드로 분리해서 처리해도 좋겠다는 생각이 들었습니다!

pkeugine replied:
음 낭비라고 하기엔 지역변수 선언이 그정도로 개선해야할만큼 무거운 작업이 아니에요.
로직을 작성할 때 변수 만드는 작업에 대해서는 성능 고려를 안하셔도 됩니다!

takoyakimchi replied:
코드가 깔끔해지기도 하고,
이름을 고민하지 않아도 된다는 장점은 확실히 있겠군요! 👍

pkeugine commented on src/main/java/validation/Validator.java:
```diff
@@ -0,0 +1,49 @@
+        }
+        validateNameDuplication(names);
+        validateCarAmount(names);
+    }
```
Validator 역시 Converter 와 비슷한 방식으로 없애거나 일부 검증 로직을 도메인쪽으로 옮겨볼 수 있을 것 같아요
예를 들어 validateCarNames 같은 경우는 현재 트레의 프로젝트 구조상 Car 도메인에게 책임을 넘겨보는거죠

Validator 를 만들어서 검증을 한 곳에 모으는 것이 절대 틀린 방식은 아니에요.
실제로 그렇게 하는 경우도 있으니까요
하지만 지금은 역할과 책임을 잘 생각해보고, 객체끼리 상호작용하는 방식으로 만들어보죠!

지금 트레가 만들어주신 애플리케이션을 보면 Validator, Service, 그리고 Converter 에 대한 테스트가 있어요
로직을 도메인 쪽으로 옮겨보면 각 도메인에 대한 단위테스트를 추가하기에도 수월할 것 같아요 :)

pkeugine commented on src/main/java/validation/Validator.java:
```diff
@@ -0,0 +1,49 @@
+        if (nonDuplicatedNames.size() != names.size()) {
+            throw new IllegalArgumentException("중복된 자동차 이름이 있습니다.");
+        }
+    }
```
Set 을 이용한 중복 검사 👍 👍 👍

pkeugine commented on src/main/java/view/IterativeReader.java:
```diff
@@ -0,0 +1,16 @@
+
+public class IterativeReader {
+
+    public static <T> T readUntilNoError(Supplier<T> supplier) {
+        while (true) {
+            try {
+                return supplier.get();
+            } catch (IllegalArgumentException e) {
+                System.out.println(e.getMessage());
+            }
+        }
+    }
```
generic, supplier 를 사용해서 자동차와 실행 횟수를 받을 때 재활용할 수 있게 했군요 👍

takoyakimchi replied:
맞습니다!
모든 프리코스 과정에서도 이와 같이 구현했습니다!
하지만 해당 메서드에서 에러 처리까지 한 번에 진행한다는 것이
역할이 너무 많아지는 것이 아닌가 싶어 마음에 걸리기도 합니다...

pkeugine replied:
괜찮다고 생각해요!

그런데 때로는 내가 하고자하는 작업에 비해 복잡해보이는 로직이 만들어지기도 해요
이 부분도 제 개인적인 생각으로는 그런 로직 중 하나라고 보는데요,
사용될 곳이 두 군데 밖에 없다면
어느정도 중복 느낌이 나는 비슷한 로직으로 그냥 만들어도 충분하다고 생각해요
개발하다보면 약간의 중복이 있어도 단순하고 쉬운 로직이 최고라고 느낄 때가 많거든요 😄 

"에러 처리하는게 역할이 너무 많아지는게 아닌가?"
이 고민은 하지 않으셔도 괜찮습니다!
"하나의 역할", "하나의 책임" 을 좀 더 너그럽게 큰 범위로 잡으셔도 돼요!

pkeugine commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,32 @@
+            System.out.println(car.getName() + " : " + "-".repeat(car.getScore()));
+        }
+        System.out.println();
+    }
```
console 출력에 알맞는 OutputView 를 만드셨네요 👍 👍

pkeugine commented on src/test/java/conversion/ConverterTest.java:
```diff
@@ -0,0 +1,55 @@
+        List<Car> carsActual = Converter.toCars(rawCarNames);
+
+        // then
+        assertEquals("aa", carsActual.get(0).getName());
```
이렇게도 써볼 수 있어요. 수정할 필요는 없습니다!
```suggestion
        assertThat(carsActual.get(0).getName()).isEqualTo("aa");
```
assertEquals() 와 assertThat().isEqualTo(), assertJ 라이브러리에 대해 한 번 알아보시고 사용해보시는 것도 추천해요!
assertJ 는 특히나 쉽고 나름 유쾌한 공식문서를 가지고 있어서, 처음으로 공식문서를 읽어보기 좋은 입문용 공식문서입니다 :)

takoyakimchi replied:
우테코 이전부터 assertJ를 사용했던 경험이 있어서
연습을 위해 JUnit을 한 번 사용해 보았습니다!
양쪽 다 더 공부해 보겠습니다!

pkeugine replied:
연습을 위해 해보는 것 좋네요!
이왕이면 미션하면서 해볼 수 있는거 다 해봐야죠~ 👍

pkeugine commented on src/test/java/conversion/ConverterTest.java:
```diff
@@ -0,0 +1,55 @@
+    @Test
+    @DisplayName("시도 횟수 변환 작업 테스트: String -> int")
+    void toRound() {
+        // given
+        String rawRound = "100";
+
+        // when
+        int roundActual = Converter.toRound(rawRound);
+
+        // then
+        assertEquals(100, roundActual);
```
BDD 방식으로 given, when, then 구분지으니 읽기 편하네요!

takoyakimchi replied:
BDD에 대해서 알지 못한 채
Spring 강의를 들었을 때 강사님이 이렇게 테스트 코드를 작성하는 것을 보고
따라서 해 보았습니다!

pkeugine commented on src/test/java/service/ServiceTest.java:
```diff
@@ -0,0 +1,34 @@
+
+        // then
+        assertEquals(new ArrayList<>(List.of("carB", "carC")), Service.getWinnerNames(cars));
+    }
```
지금 이 테스트는 Car 객체가 `move()` 를 잘 수행할거라고 믿고 실행하고 있어요.
그렇다면 Car 클래스의 `move()` 관련 테스트도 있으면 좋겠네요!

takoyakimchi replied:
테스트 작성 과정에서 같은 생각을 했었고
`move()`에서 단순 increment만 하기 때문에 테스트 코드 작성을 생략했었는데
안일한 생각이었던 것 같습니다!

pkeugine replied:
협업하는 팀원이 있었다면, 그 사람은 트레가 작업한 내용을 잘 알지 못할거에요. 얼마나 간단한지, 얼마나 복잡한지 모르는거죠
그럴 때 가장 믿을 수 있는게 바로 테스트라고 생각합니다 :)

takoyakimchi replied:
로직이 간단해도 협업하는 팀원은 알지 못할 수 있겠군요!

테스트를 짜는 범위를 더 넓혀 보도록 하겠습니다! :)

pkeugine commented on src/test/java/validation/ValidatorTest.java:
```diff
@@ -0,0 +1,86 @@
+        List<String> nameWithDuplication = new ArrayList<>(List.of("aa,bb,bb,cc"));
+
+        // when - then
+        assertThrows(IllegalArgumentException.class,
+            () -> Validator.validateCarNames(nameTooLong));
+        assertThrows(IllegalArgumentException.class,
+            () -> Validator.validateCarNames(nameBlank));
+        assertThrows(IllegalArgumentException.class,
+            () -> Validator.validateCarNames(nameWithIllegalLetter));
+        assertThrows(IllegalArgumentException.class,
+            () -> Validator.validateCarNames(nameWithDuplication));
```
이렇게 여러 assert 문이 하나의 테스트 메서드에 있을 때, `assertAll()` 안에 넣어줄 수도 있어요
이러면 어떤 차이가 있을까요?

takoyakimchi replied:
지금의 방식대로라면 중간에서 실패가 뜨는 경우 테스트를 멈춰 버려, 실행되지 않은 다른 테스트의 성패 여부를 알 수 없는데
`assertAll()`을 사용하면 실패가 뜨더라도, 다른 테스트의 성패 여부를 알 수 있습니다!

`assertAll()` 활용해 보겠습니다.

pkeugine replied:
💯

pkeugine commented on src/test/java/validation/ValidatorTest.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    @Test
+    @DisplayName("차 이름 테스트: 예외 입력(2) - 차가 100대 넘는 경우")
```
차가 100대 넘는 경우 에러가 나도록 한 추가구현 잘 만드셨어요 👍 
그런데 이 테스트를 보고 두 가지 궁금증이 생겼습니다! : 
1. 그렇다면 최소 1 대는 있어야한다는 검증도 테스트해보면 좋을 것 같다
2. '100대 넘는 경우' 를 검증하는데, 100 이라는 구간은 테스트하지 않고 좀 더 높은 값을 만들어 테스트한다

테스트 케이스를 추가하고 100이라는 값에 근접하도록 테스트를 수정해보는건 어떨까요?

takoyakimchi replied:
101대 테스트 케이스를 만드는 것이 번거로운 작업이다보니
경계값을 생각 안하고 테스트를 짰던 것 같습니다

경계값에 대한 테스트 진행하도록 하겠습니다!

pkeugine commented on src/test/java/validation/ValidatorTest.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    @Test
+    @DisplayName("라운드 테스트 : 예외입력")
```
이전 테스트는 제목에 구체적인 상황을 적어주셨어요
이 테스트도 그렇게 구체적으로 적어주면 어떨까요?

takoyakimchi replied:
페어 프로그래밍 과정에서 네이밍을 맞추는 작업이 부족했던 것 같습니다
DisplayName 컨벤션 맞추도록 하겠습니다!

pkeugine reviewed (changes requested):
안녕하세요 트레!
제가 새롭게 알려드릴 점이 없어보여요!

static 클래스에 많이 의존하는 부분을 수정하고 자잘한 테스트 수정하시면 될 것 같네요
이미 자바의 많은 부분을 알고 계신 것 같지만, 그래도 궁금하신 점이나 같이 이야기해보고 싶은 점이 있다면 언제든지 DM 주세요 :)

그리고 PR 올리실 때도 어떤 고민을 하셨는지 적어주시면 같이 이야기할 거리도 많아져요! 적극 활용해보시는걸 추천합니다 👍

takoyakimchi committed:
style: 모든 파일 마지막에 new line 추가

takoyakimchi committed:
style: unused import 제거

takoyakimchi committed:
refactor: getter가 아닌 메서드의 이름 변경

getCars -> findCars
getRound -> findRound

takoyakimchi committed:
refactor: controller 삭제하고 해당 부분 main으로 이동

takoyakimchi committed:
feat(error): 에러 처리 클래스 작성 - 에러 발생할 시 해당 부분부터 재실행

takoyakimchi committed:
refactor: Service, Controller, Validator, Converter를 사용하지 않는 방식으로 변경

Service에 있던 비즈니스 로직을 RacingGame 도메인에 구현
Controller에 있던 게임 시작을 Main으로 이동
Validator에 있던 검증 로직을 도메인으로 이동
Converter에 있던 변환 로직을 도메인으로 이동
Round 도메인 추가

takoyakimchi committed:
docs: TODO 일부 작성

takoyakimchi committed:
refactor(error): 메서드 이름 변경 및 Runnable -> Supplier 변경

리턴 값 필요에 따른 함수형 인터페이스 변경

takoyakimchi committed:
refactor: 예외 발생시 재입력 구현

takoyakimchi committed:
style: 예외 발생시 재입력 구현(주석 제거)

takoyakimchi committed:
refactor: 게임 플레이에 대한 비즈니스 로직을 도메인으로 이동

takoyakimchi committed:
refactor(racing-game): play 메서드 내 메서드 분리를 통한 depth 최소화

takoyakimchi committed:
refactor: 움직임에 대한 조건을 담당하는 인터페이스 생성 (책임 분리)

takoyakimchi committed:
refactor: 랜덤 관련 상수를 실제로 사용하는 곳으로 이동 & 상수명을 설계의도에 맞게 변경

takoyakimchi committed:
style: 해결된 TODO 주석 삭제

takoyakimchi committed:
refactor: 메시지 상수를 담던 클래스 제거, 해당 상수는 실제 사용되는 클래스로 이동

takoyakimchi committed:
refactor: 사용하지 않는 파일 제거

view/IterativeReader.java 제거

takoyakimchi committed:
style: interface 내부 첫 줄에 newline 추가

takoyakimchi committed:
style: interface 내부 첫 줄에 newline 추가

takoyakimchi committed:
refactor(view): 변수명에서 type명을 사용하지 않도록 변경

takoyakimchi committed:
refactor(racing-game): 접근 제어자 변경, 메서드 순서 정리

같은 클래스에서만 쓰이는 static 메서드를 private으로 변경, 메서드 순서를 사용되는 순서대로 정렬

takoyakimchi committed:
refactor: static일 필요가 없는 메서드에서 static을 제거

takoyakimchi committed:
refactor: 불필요한 import문 제거

takoyakimchi committed:
refactor: 불필요한 try-catch 제거

takoyakimchi committed:
refactor: 예외 처리 로직 수정

"a,b,c,d,,," 입력 시 정상 처리하던 오류 수정

takoyakimchi committed:
refactor: 불필요한 주석 제거

takoyakimchi committed:
refactor: 상수명 의미에 맞게 변경

"차의 대수" == "car amount"

takoyakimchi committed:
refactor: 메서드명을 자연어와 비슷하게 보이도록 변경

make round from (raw round)

takoyakimchi committed:
refactor: 도메인이 출력을 담당하지 않도록 변경

takoyakimchi committed:
refactor: Round의 validation 로직 정리

takoyakimchi committed:
refactor: MoveStrategy를 RacingGame으로 이동

takoyakimchi committed:
feat(test): Car 도메인 단위 테스트 추가

takoyakimchi committed:
feat(test): RacingGame 도메인 단위 테스트 추가

takoyakimchi committed:
feat(test): Round 도메인 단위 테스트 추가

takoyakimchi committed:
refactor(test): 여러 assert문이 있는 테스트에 대해 assertAll 적용

takoyakimchi committed:
refactor: 클래스 내에서만 사용하는 상수의 접근제어자를 private으로 변경

takoyakimchi committed:
refactor: magic number 정의를 실제 사용되는 클래스로 이동

takoyakimchi committed:
refactor: 글자수 제한 매직넘버화

takoyakimchi committed:
refactor: static 클래스 최대한 제거

takoyakimchi committed:
refactor: public 메서드가 위로 오도록 수정

takoyakimchi committed:
refactor: domain이 구현체에 의존하지 않도록 수정

takoyakimchi commented:
안녕하세요! 리뷰 받고 많은 도움이 되었습니다. 감사합니다!!!
덕분에 보람차게 과제를 진행할 수 있었습니다.
아직 1주차인데도 얻어가는 것이 굉장히 많은 것 같아 기분이 좋습니다 😆

말씀해 주신 부분 반영하여 전체적인 구조를 수정해 보았습니다.

## 변경 사항
- `Messages` 클래스를 삭제하고 `InputView` 와 `OutputView` 로 해당 부분을 이동하였습니다.
- `Controller` 클래스를 삭제하고 해당 부분을 `Main` 으로 이동하였습니다.
- 과한 클래스 분리를 해소하고자 노력했습니다.
    - `Service` `Converter` `Validator` 에 있던 로직을 도메인에서 처리하도록 하였습니다.
    - 이에 따라서 `Car`  도메인 이외에 `RacingGame` `Round` 도메인을 추가하였습니다.
- `IterativeReader` 클래스는 reader의 역할보다는 에러 처리의 역할이 더 강하다고 생각하여
`error/ErrorHandler` 로 따로 분리하였으며, 메서드명도 역할에 맞게 변경하였습니다.
- 우승자 판별 로직을 변경하였습니다.
    - `Car` 에 `compareTo()` `equals()`를 오버라이딩하였고 이것을 기반으로 우승자를 판별할 수 있도록 하였습니다.
- 전진 조건에 대한 부분을 `MoveStrategy` 로 인터페이스화 하였으며, `RandomPowerMoveStrategy` 라는 클래스가
이를 구현하도록 하여, 전진 조건에 대한 확장성을 높였습니다.
- 테스트 코드에서 `assertAll()` 을 적용해 보았습니다.
- `static` 사용을 최대한 줄이고자 노력했습니다.

## 의문점
- 현재 코드에서는 `InputView.readCarNames()`와 `InputView.readRound()`에서 객체를 생성하고 있는데,
이러한 구조가 바람직한지 궁금합니다!
- 입력 및 예외처리를 수월하게 하기 위해 `Round` 도메인을 추가하였는데, 의미상으로는 `RacingGame` 안에
`int round`라는 필드를 추가하는 것이 더 바람직하지 않나 하는 의문도 듭니다.
이것에 대한 피케이의 생각이 궁금합니다!

pkeugine commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,23 @@
+import view.InputView;
+import view.OutputView;
+
+public class Main {
+
+    public static void main(String[] args) {
+        ErrorHandler errorHandler = new ErrorHandler();
+        InputView inputView = new InputView();
+        OutputView outputView = new OutputView();
+
+        RacingGame racingGame = errorHandler.executeErrorOccurable(inputView::readCarNames);
+        Round round = errorHandler.executeErrorOccurable(inputView::readRound);
+
+        System.out.println("실행 결과");
+        for (int i = 0; i < round.getRound(); i++) {
+            outputView.printScore(racingGame.playOneRound());
+        }
+        outputView.printCars(racingGame.findWinners());
+    }
+}
```
Controller 를 제거하고 Main 으로 로직을 옮긴점 💯💯💯💯💯💯💯
Service 와 Controller 가 사라지면서 구조가 훨씬 깔끔해졌죠!

간단하고 깔끔한게... 화려하고 클래스가 많은 것보다 절대 나쁜게 아니니까요!

pkeugine commented on src/main/java/domain/Car.java:
```diff
@@ -0,0 +1,49 @@
+        this.score++;
+    }
+
+    @Override
+    public int compareTo(Car otherCar) {
+        return Integer.compare(this.score, otherCar.score);
+    }
+
+    @Override
+    public boolean equals(Object otherCar) {
+        if (!(otherCar instanceof Car)) {
+            throw new RuntimeException("알 수 없는 예외가 발생했습니다.");
+        }
+
+        return this.score == ((Car) otherCar).score;
+    }
+
+    private void validate(String name) {
+        if (!name.matches("^[a-zA-Z]{" + MIN_NAME_LENGTH + "," + MAX_NAME_LENGTH + "}$")) {
+            throw new IllegalArgumentException(
+                MIN_NAME_LENGTH + "~" + MAX_NAME_LENGTH + "자의 알파벳만 입력 가능합니다.");
+        }
+    }
+}
```
검증로직이 Car 쪽으로 넘기기 👍 

저번 리뷰에서 메서드의 순서에 대해 대화했었죠?
트레의 메서드 순서 기준이 있다는 점이 좋았습니다!
그런데 주변의 다른 분들이 보통 어떤 순서대로 두는지도 물어본 뒤,
어느정도 대중적인 방식을 따라도 괜찮을 것 같아요

맞고 틀린건 없지만
보통 비즈니스 로직, 검증로직처럼 자주 보고 자주 변경할 것들부터 위에 적거든요!

takoyakimchi replied:
public 메서드는 위에 둬야만 한다! 라는 강박이 있었던 것 같습니다!
순서 알아보고 적용시켜 보겠습니다!

pkeugine commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,23 @@
+import view.InputView;
+import view.OutputView;
+
+public class Main {
+
+    public static void main(String[] args) {
+        ErrorHandler errorHandler = new ErrorHandler();
+        InputView inputView = new InputView();
+        OutputView outputView = new OutputView();
```
커밋을 보니 TODO 를 남겨주셨던 흔적이 보이네요 269fa0bf
리팩토링하면서 다 지운 점이 좋은것 같아요 👍 
커밋을 작게 분리하는 것은 좋지만, 굳이 TODO 만을 위한 커밋을 만들 필요는 없어요 :)

takoyakimchi replied:
> 굳이 TODO 만을 위한 커밋을 만들 필요는 없어요 :)

이것도 필요성에 대한 고민을 한 적이 있었는데
시원하게 해결되었습니다. 감사합니다!

pkeugine commented on src/main/java/error/ErrorHandler.java:
```diff
@@ -0,0 +1,18 @@
+
+import java.util.function.Supplier;
+
+public class ErrorHandler {
+
+    private static final String ERROR_PREFIX = "[ERROR] ";
+
+    public <T> T executeErrorOccurable(Supplier<T> supplier) {
```
"입력을 받은 뒤, 유효하지 않다면 유효할 때까지 다시 입력 받는다."

이걸 간단하고 잘 읽히게 표현할 이름으로 바꿔보는건 어떨까요?
지금은 메서드명으로 저 작업을 유추해내기 좀 어려울 것 같아요
이전의 IterativeReader 가 조금 더 명확하지 않았나 싶기도 하구요

takoyakimchi replied:
메서드명에 꽤나 신경을 쓰고 있다고 생각했는데
더 좋은 이름 생각해 보겠습니다!

pkeugine commented on src/main/java/domain/RacingGame.java:
```diff
@@ -0,0 +1,66 @@
+    private final List<Car> cars;
+    private final MoveStrategy moveStrategy;
+
+    public RacingGame(String rawCarNames, MoveStrategy moveStrategy) {
+        cars = new ArrayList<>();
+        this.moveStrategy = moveStrategy;
+        makeCarsFrom(rawCarNames);
+    }
+
+    public List<Car> playOneRound() {
+        for (Car car : cars) {
+            if (moveStrategy.isMovable()) {
+                car.move();
+            }
+        }
+        return cars;
+    }
+
+    public List<Car> findWinners() {
+        return cars.stream()
+            .filter(car -> car.equals(Collections.max(cars)))
+            .toList();
+    }
```
오 public 한 메서드를 가장 위에 두고,
private 한 것들을 아래에 두셨군요!
저도 이렇게 해요 😉

pkeugine commented on src/main/java/movestrategy/RandomPowerMoveStrategy.java:
```diff
@@ -0,0 +1,16 @@
+
+public class RandomPowerMoveStrategy implements MoveStrategy {
+
+    private static final int MAXIMUM_POWER = 10;
+    private static final int MINIMUM_POWER_TO_MOVE = 4;
```
상수 이름 정말 잘 바꾸셨네요 👍 
이전보다 훨씬 명확해졌어요

pkeugine commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,27 @@
+
+public class InputView {
+
+    private static final String INPUT_CAR_NAMES = "경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).";
+    private static final String INPUT_ROUND = "시도할 회수는 몇회인가요?";
```
메시지에 사용되는 문자열을 상수로 만들면 얻는 이점과 잃은 것이 뭐가 있을까요? 🤔

takoyakimchi replied:
클래스의 상단에 위치하기 때문에 수정에 용이할 것 같습니다.

하지만 `System.out.println()` 내에 메시지 문자열이 직접 위치하는 것보다
코드를 읽을 때 가독성이 떨어지지 않을까 생각합니다.

pkeugine commented on src/main/java/domain/RacingGame.java:
```diff
@@ -0,0 +1,66 @@
+    private final List<Car> cars;
+    private final MoveStrategy moveStrategy;
+
+    public RacingGame(String rawCarNames, MoveStrategy moveStrategy) {
+        cars = new ArrayList<>();
+        this.moveStrategy = moveStrategy;
+        makeCarsFrom(rawCarNames);
+    }
```
cars 가 지금은 그렇지 않지만, 생성자에서는 파라미터와 필드명이 겹칠 때가 많아요. `moveStrategy` 처럼요.
그래서 필드는 `this` 를 붙여서 명확하게 하는 습관을 들여봐도 좋을 것 같아요

takoyakimchi replied:
`this`를 작성하면 더 깔끔하고 명확해지겠네요!

pkeugine commented on src/test/java/domain/CarTest.java:
```diff
@@ -0,0 +1,105 @@
+        );
+    }
+
+    @Test
+    @DisplayName("이름 예외 입력: 자동차의 이름은 공백일 수 없다.")
+    void carName_exception_noBlank_noEmpty() {
+        Assertions.assertAll(
+            () -> assertThatThrownBy(() -> new Car(""))
+                .isInstanceOf(IllegalArgumentException.class),
+            () -> assertThatThrownBy(() -> new Car(" "))
+                .isInstanceOf(IllegalArgumentException.class)
+        );
+    }
+
+    @Test
+    @DisplayName("이름 예외 입력: 자동차의 이름은 1~5자만 허용된다.")
+    void carName_exception_cantViolateLengthRegulations() {
+        Assertions.assertAll(
+            () -> assertThatThrownBy(() -> new Car(""))
+                .isInstanceOf(IllegalArgumentException.class),
+            () -> assertThatThrownBy(() -> new Car("abcdef"))
+                .isInstanceOf(IllegalArgumentException.class)
+        );
+    }
```
저번 리뷰에서 같이 언급했던 assertAll 사용 좋네요 👍👍👍👍
더 나아가 `@ParameterizedTest`  어노테이션도 한 번 학습해보시죠!

pkeugine commented on src/test/java/domain/RacingGameTest.java:
```diff
@@ -0,0 +1,101 @@
+        );
+    }
+
+    @Test
+    @DisplayName("자동차 이름 정상 입력: 경계값 입력 - 자동차 100대")
+    void carNames_ok_boundaryTest_hundredCases() {
+        new RacingGame(
+            "a    ,b    ,c    ,d    ,e    ,f    ,g    ,h    ,i    ,j    ,"
+                + "k    ,l    ,m    ,n    ,o    ,p    ,q    ,r    ,s    ,t    ,"
+                + "u    ,v    ,w    ,x    ,y    ,z    ,aa   ,bb   ,cc   ,dd   ,"
+                + "ee   ,ff   ,gg   ,hh   ,ii   ,jj   ,kk   ,ll   ,mm   ,nn   ,"
+                + "oo   ,pp   ,qq   ,rr   ,ss   ,tt   ,uu   ,vv   ,ww   ,xx   ,"
+                + "yy   ,zz   ,aaa  ,bbb  ,ccc  ,ddd  ,eee  ,fff  ,ggg  ,hhh  ,"
+                + "iii  ,jjj  ,kkk  ,lll  ,mmm  ,nnn  ,ooo  ,ppp  ,qqq  ,rrr  ,"
+                + "sss  ,ttt  ,uuu  ,vvv  ,www  ,xxx  ,yyy  ,zzz  ,aaaa ,bbbb ,"
+                + "cccc ,dddd ,eeee ,ffff ,gggg ,hhhh ,iiii ,jjjj ,kkkk ,llll ,"
+                + "mmmm ,nnnn ,oooo ,pppp ,qqqq ,rrrr ,ssss ,tttt ,uuuu ,vvvv"
+            , moveStrategy);
+    }
```
존경합니다 트레...
저보고 이름 100 개 관련 테스트를 짜보라고 한다면... 이렇게 짤 생각을 못했을텐데...
보니까 나름 직관적이고 좋네요 🤔 
하지만 경계값이 1000, 10000 이라도 이렇게 할 수 있을지, 경계값이 자주 바뀌면 이걸 매번 수정할지 고민해보죠! 😄

takoyakimchi replied:
서로 다른 이름을 가지는 문자열 n개를(규칙에 맞게) 만드는 테스트용 스태틱 메서드를
테스트 클래스 내에 만들어 주어도 되겠군요!

pkeugine commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,23 @@
+        RacingGame racingGame = errorHandler.executeErrorOccurable(inputView::readCarNames);
+        Round round = errorHandler.executeErrorOccurable(inputView::readRound);
+
+        System.out.println("실행 결과");
```
이건 OutputView 에 있을 법 한 녀석이네요!

pkeugine commented on src/main/java/movestrategy/MoveStrategy.java:
```diff
@@ -0,0 +1,6 @@
+package movestrategy;
+
+public interface MoveStrategy {
+
+    boolean isMovable();
+}
```
MoveStrategy 를 인터페이스로 추출한 이유가 있을까요?
RandomPowerMoveStrategy 라는 구현체 하나 뿐이라서요!

pkeugine replied:
"구현체가 하나만 있는 인터페이스는 굳이 인터페이스화 할 필요가 있을까?"
vs
"유연하고 확장 가능한 구조를 위해 구현체가 하나만 있더라도 인터페이스가 필요해"

전략 패턴이라는 방식을 잠시 잊고 고민해봐도 좋을 것 같아요

takoyakimchi replied:
"전진 조건"을 의미상으로 생각해 보았을 때 언제든지 로직이 추가되거나 변경될 수도 있겠다는 생각을 하였고,

> "유연하고 확장 가능한 구조를 위해 구현체가 하나만 있더라도 인터페이스가 필요해"

쪽으로 생각하여 이번에는 인터페이스로 구현하였습니다!

pkeugine commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,25 @@
+
+    private static final String WIN_THE_GAME = "가 최종 우승했습니다.";
+
+    public void printScore(List<Car> cars) {
+        for (Car car : cars) {
+            System.out.println(car.getName() + " : " + "-".repeat(car.getScore()));
+        }
+        System.out.println();
+    }
+
+    public void printCars(List<Car> cars) {
+        List<String> carNames = new ArrayList<>();
+        for (Car car : cars) {
+            carNames.add(car.getName());
+        }
+        System.out.println(String.join(", ", carNames) + WIN_THE_GAME);
```
전체적으로 OutputView 쪽에 stream 을 활용해볼수도 있겠네요
예를 들어 `printCars` 쪽에서 `carNames` 를 만들 때 아래와 같이 for 문을 stream 으로 변경해볼 수 있어요:
```java
  public void printCars(List<Car> cars) {
    List<String> carNames = cars.stream()
        .map(Car::getName)
        .toList();
    System.out.println(String.join(", ", carNames) + WIN_THE_GAME);
  }
```

또는, 아래의 `String.join(", ", carNames)` 부분까지 처리할 수도 있어요:
```java
  public void printCars(List<Car> cars) {
    String result = cars.stream()
        .map(Car::getName)
        .collect(Collectors.joining(", "));
    System.out.println(result + WIN_THE_GAME);
  }
```

pkeugine commented on src/test/java/domain/CarTest.java:
```diff
@@ -0,0 +1,105 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class CarTest {
```
테스트 정말 꼼꼼하게 작성하셨네요!!! 👍👍👍 

그리고 테스트 이름도 정말 섬세하고 좋네요

pkeugine reviewed (commented):
안녕하세요 트레!
놀라울정도로 리뷰도 잘 반영해주셨고, 테스트도 더 철저하게 작성해주셨네요
트레가 적어주신 변경사항과 남겨주신 커밋 모두 다 확인해봤습니다
열심히 해주셔서 저도 리뷰하면서 재밌었네요 :)

트레의 의문점에 대한 제 답변을 남겨볼게요 : 

* `InputView` 에서 객체를 생성하는 것이 바람직한가

이 부분은 크게 걱정할 요소는 아니라고 봐요
`InputView` 에서 그냥 입력값만 받아서 Main 쪽에서 객체를 생성하던지, 아니면 트레처럼 만들던지
가장 중요한건 결국 "그 입력값에 대한 검증을 했는가?" 라고 저는 생각하거든요
어차피 `OutputView` 에서도 지금 도메인을 받아서 출력값을 만들어주고 있잖아요? :)

* Round 객체 vs int round

제 생각을 물어보셨으니... 저라면 굳이 객체를 만들지 않고 int round 값으로 다룰거에요.
round 는 양수이기만 해도 되는 값이라 검증할 부분이 매우 적고, 딱히 round 만의 로직이라고 할 게 없거든요
그래서 저라면 객체로 만들지 않을거에요 :)

pkeugine reviewed (approved):
고생하셨어요 트레!
몰라볼 정도로 리팩토링 하셨네요!
훨씬 깔끔하게 변경되었어요 👍 
이번 step 은 그 리팩토링만으로도 충분한거 같으니
이제 다음 step 으로 넘어가보죠

이번에 남긴 리뷰를 반영하신다면 step 2 에 포함시켜주세요!
제가 step 2 에서 확인해볼게요
조금만 더 파이팅해보죠!!!

궁금한 점이나 더 대화하고 싶은 것이 있다면 언제든지 DM 주세요!!!
